### PR TITLE
Allow to use any exception handler as the formatter

### PR DESCRIPTION
**Pull request by AlexMasterov**
My attempt to remake the (Whoops) [`ExceptionHandler`](https://github.com/equip/framework/blob/e7a51ea086a2ec9ec0fae6766ebfc6c72f883694/src/Handler/ExceptionHandler.php). 
These changes have a huge impact on performance. (less reponse time) 
But it's still work in progress, I have a proposal that may affect the final result.

I used the Whoops handlers as a formatters, which are used a generic Responder like [`FormattedResponder`](https://github.com/equip/framework/blob/e618b83ff90ffcac0fd6b28a915d7119e40efbbb/src/Responder/FormattedResponder.php) and DTO [`Payload`](https://github.com/equip/framework/blob/37c10f1aaa0328954011c18cdc1cbd66a4a7d8f2/src/Payload.php). That seems the obvious way, and does not violate the ADR pattern:

[Action-Domain-Responder: View vs Responder](https://github.com/pmjones/adr#view-vs-responder)

> **Note also that a generic Responder may be used by more than one Action.** 
> The point is that the Action leaves all header and content work to the Responder, not that there must be a different Responder for each different View.

**My suggestions**
To make an interfaces for exceptions, it's a sketch:

``` php
use Equip\Exception\ExceptionInterface;
interface EquipExceptionHandlerInterface
{
    public function __invoke(
        ServerRequestInterface $request,
        ResponseInterface $response,
        callable $next
    );
    public function withEquipException(ExceptionInterface $exception);
}

use Exception;
interface ExceptionHandlerInterface
{
    public function withException(Exception $exception);
}

use Throwable;
interface ThrowableHandlerInterface
{
    public function withThrowable(Throwable $throwable);
}
```

Use a PSR logger as a decorator. Soon I'll add it.
Use the Whoops as a separate package. It already looks like that.
Use for the Equip a simple (native) exception handler by default, and change it using a DI.

``` php
use Equip\Exception\ExceptionInterface;
use Exception;
public function __invoke(
    ServerRequestInterface $request,
    ResponseInterface $response,
    callable $next
) {
    try {
        return $next($request, $response);
    } catch (ExceptionInterface $exception) {
        return $this->exceptionHandler->withEquipException($request, $response, $exception);
    } catch (Exception $exception) {
        return $this->exceptionHandler->withException($request, $response, $exception);
    }
}
```

### CODE REVIEW

**Comment by AlexMasterov** [comment]
Some of the suggestions is revoking, please review the current version. The old tests are working, but they still need to improve and to add more cases.

File: src/Formatter/WhoopsHtmlFormatter.php [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+     * @var Whoops
+     */
+    private $whoops;
+
+    /**
+     * @param Whoops      $whoops
+     * @param HtmlHandler $handler
+     */
+    public function __construct(
+        Whoops      $whoops,
```
Comment (ameech) [inline src/Formatter/WhoopsHtmlFormatter.php]: These don't need to be aligned.

File: src/Formatter/WhoopsJsonFormatter.php [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+     * @var Whoops
+     */
+    private $whoops;
+
+    /**
+     * @param Whoops      $whoops
+     * @param JsonHandler $handler
+     */
+    public function __construct(
+        Whoops      $whoops,
```
Comment (ameech) [inline src/Formatter/WhoopsJsonFormatter.php]: Here too and so on.

File: src/Handler/ExceptionHandler.php [hunk 3]
```diff
@@ -2,68 +2,65 @@
...
+    use ResolverTrait;
 
     /**
-     * @var ExceptionHandlerPreferences
+     * Allowed range for a valid HTTP status code
      */
-    private $preferences;
+    const MINIMUM = 100;
+    const MAXIMUM = 599;
+    const FAULTY = 500;
```
Comment (ameech) [inline src/Handler/ExceptionHandler.php]: What is `FAULTY`?

File: src/Handler/ExceptionHandler.php [hunk 4]
```diff
@@ -74,77 +71,112 @@ public function __invoke(
...
-     * @param string $type
+     * @return ResponseInterface
+     */
+    private function format(
+        ServerRequestInterface $request,
+        ResponseInterface $response,
+        Exception $exception
+    ) {
+        $formatter = $this->formattedWithFormatters($this->formatters);
+        $payload = $this->payloadWithOutput([
```
Comment (ameech) [inline src/Handler/ExceptionHandler.php]:
```PHP
$payload = $this->payloadWithOutput(compact('exception'))
```

File: tests/Formatter/JsonFormatterTest.php [hunk 5]
```diff
@@ -20,7 +20,15 @@ protected function setUp()
 
     public function testAccepts()
     {
-        $this->assertEquals(['application/json'], JsonFormatter::accepts());
+        $accepts = [
+            'application/json',
+            'application/ld+json',
+            'application/vnd.api+json',
+            'application/vnd.geo+json',
+            'application/javascript'
```
Comment (ameech) [inline tests/Formatter/JsonFormatterTest.php]: Missing `,`.

File: tests/Handler/ExceptionHandlerTest.php [hunk 6]
```diff
@@ -39,15 +37,19 @@ public function dataTypes()
...
-            ['application/json'],
-            ['application/ld+json'],
-            ['application/vnd.api+json'],
-            ['application/vnd.geo+json'],
-            ['application/xml'],
-            ['application/atom+xml'],
-            ['application/rss+xml'],
-            ['text/plain'],
+            [
+                'application/json',
```
Comment (ameech) [inline tests/Handler/ExceptionHandlerTest.php]: `JsonFormatter::accepts()` so we're not have to duplicate data.

File: tests/Handler/ExceptionHandlerTest.php [hunk 7]
```diff
@@ -39,15 +37,19 @@ public function dataTypes()
...
+                'application/ld+json',
+                'application/vnd.api+json',
+                'application/vnd.geo+json',
+                'application/javascript'
+            ],
+            [
+                'application/xml',
```
Comment (ameech) [inline tests/Handler/ExceptionHandlerTest.php]: Same here.

File: src/Handler/ExceptionHandler.php (hunk 4 above)
Comment (AlexMasterov) [inline src/Handler/ExceptionHandler.php]: @ameech I forgot about it. Thank you.

File: src/Formatter/WhoopsHtmlFormatter.php [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+class WhoopsHtmlFormatter extends HtmlFormatter
+{
+    /**
+     * @var Whoops
+     */
+    private $whoops;
+
+    /**
+     * @param Whoops      $whoops
```
Comment (ameech) [inline src/Formatter/WhoopsHtmlFormatter.php]: Nitpicking now, but don't need to align these either. Just makes diffs messy.

File: src/Handler/ExceptionHandler.php [hunk 9]
```diff
@@ -74,77 +71,110 @@ public function __invoke(
...
+     *
+     * @return ResponseInterface
+     */
+    private function status(
+        ResponseInterface $response,
+        Exception $exception
+    ) {
+        $exceptionCode = $exception->getCode();
+
+        $code = filter_var($exceptionCode, FILTER_VALIDATE_INT, ['options' => [
```
Comment (ameech) [inline src/Handler/ExceptionHandler.php]:
```PHP
$code = filter_var($exceptionCode, FILTER_VALIDATE_INT, [
  'options' => [
    ..
  ],
]);
```

File: src/Formatter/HtmlFormatter.php [hunk 10]
```diff
@@ -3,6 +3,7 @@
 namespace Equip\Formatter;
 
 use Equip\Adr\PayloadInterface;
+use Equip\Formatter\FormatterInterface;
```
Comment (ameech) [inline src/Formatter/HtmlFormatter.php]: Shouldn't have to add this.

File: src/Handler/ExceptionHandler.php (hunk 3 above)
Comment (AlexMasterov) [inline src/Handler/ExceptionHandler.php]: @ameech This code is used when the HTTP status code is invalid. (old case)

File: src/Formatter/WhoopsHtmlFormatter.php (hunk 8 above)
Comment (AlexMasterov) [inline src/Formatter/WhoopsHtmlFormatter.php]: @ameech I know about it. But in this project I see different styles, so I thought that this is not so important. It would be nice to finally decide.

**Comment by AlexMasterov** [comment]
@ameech thank you for the review. The most important thing for me that you have corrected (project) code style. Old tests need to improve, I wrote about this.

**Comment by AlexMasterov** [comment]
@ameech please review the current version handlers [`ExceptionHandler`](https://github.com/AlexMasterov/framework/blob/0c190e960a8d4e0102d5336c9edf515be042cb52/src/Handler/ExceptionHandler.php) and [`ExceptionLoggerHandler`](https://github.com/AlexMasterov/framework/blob/0c190e960a8d4e0102d5336c9edf515be042cb52/src/Handler/ExceptionLoggerHandler.php).

File: src/Exception/HttpException.php [hunk 11]
```diff
@@ -54,21 +61,15 @@ public static function badRequest($message)
...
-
-    /**
-     * @param ResponseInterface $response
+     * @param string $path
      *
-     * @return ResponseInterface
+     * @return static
      */
-    public function withResponse(ResponseInterface $response)
+    public static function notFound($path)
```
Comment (shadowhand) [inline src/Exception/HttpException.php]: `static` functions should be placed before class methods.

File: src/Formatter/HtmlFormatter.php [hunk 12]
```diff
@@ -2,9 +2,7 @@
 
 namespace Equip\Formatter;
 
-use Equip\Adr\PayloadInterface;
-
-abstract class HtmlFormatter implements FormatterInterface
```
Comment (shadowhand) [inline src/Formatter/HtmlFormatter.php]: Why was this changed?

File: src/Formatter/PlatesFormatter.php [hunk 13]
```diff
@@ -3,10 +3,11 @@
 namespace Equip\Formatter;
 
 use Equip\Adr\PayloadInterface;
+use Equip\Formatter\FormatterInterface;
 use Equip\Formatter\HtmlFormatter;
 use League\Plates\Engine;
 
-class PlatesFormatter extends HtmlFormatter
+class PlatesFormatter extends HtmlFormatter implements FormatterInterface
```
Comment (shadowhand) [inline src/Formatter/PlatesFormatter.php]: I don't see what benefit is gained by pushing the `implements` call to extensions of `HtmlFormatter`. It duplicates code unnecessarily.

File: src/Formatter/WhoopsXmlFormatter.php [hunk 14]
```diff
@@ -0,0 +1,45 @@
+<?php
+
+namespace Equip\Formatter;
+
+use Equip\Formatter\FormatterInterface;
+use Equip\Formatter\WhoopsRenderTrait;
+use Whoops\Handler\XmlResponseHandler as XmlHandler;
+use Whoops\Run as Whoops;
+
+class WhoopsXmlFormatter implements FormatterInterface
```
Comment (shadowhand) [inline src/Formatter/WhoopsXmlFormatter.php]: I don't want to include support for XML formatting. No one should need that these days.

File: src/Handler/ExceptionHandler.php [hunk 15]
```diff
@@ -2,62 +2,64 @@
...
+use Equip\Responder\FormattedResponder;
 use Exception;
-use InvalidArgumentException;
-use Negotiation\Negotiator;
 use Psr\Http\Message\ResponseInterface;
 use Psr\Http\Message\ServerRequestInterface;
-use Psr\Log\LoggerInterface;
 use Relay\ResolverInterface;
-use Whoops\Run as Whoops;
+use Throwable;
```
Comment (shadowhand) [inline src/Handler/ExceptionHandler.php]: Isn't this PHP7+ only?

**Comment by shadowhand** [comment]
I like the idea of using formatters for exceptions. Unfortunately, this PR does way too many things at once. The modifications to extension/implements need to be removed before this can be merged.

As a general rule of thumb, a PR should either: a) add a feature, b) fix a bug, or c) refactor. PRs that  do multiple things at once are hard to review and make the history harder to follow.

File: src/Handler/ExceptionHandler.php (hunk 15 above)
Comment (AlexMasterov) [inline src/Handler/ExceptionHandler.php]: No, and caught by type hinting, but is used without. (it's safe)

File: src/Formatter/HtmlFormatter.php (hunk 12 above)
Comment (AlexMasterov) [inline src/Formatter/HtmlFormatter.php]: Ref: [comment](https://github.com/equip/framework/pull/56#issuecomment-223758723).

Maybe I misunderstood.

Comment (shadowhand) [inline src/Formatter/HtmlFormatter.php]: This change should be reverted.

**Comment by AlexMasterov** [comment]
@shadowhand Please check the PR now.

**Comment by AlexMasterov** [comment]
@shadowhand what do I need to do to get this merged?

I'll open at least two PR which depend on these changes:
- ExceptionLoggerHandler (decorator)
- Implementation of ExceptionInterface (marker type) for exceptions coming from the Equip

File: src/Exception/ExceptionInterface.php [hunk 16]
```diff
@@ -0,0 +1,7 @@
+<?php
+
+namespace Equip\Exception;
+
+interface ExceptionInterface
```
Comment (shadowhand) [inline src/Exception/ExceptionInterface.php]: This doesn't make sense to me. We shouldn't need to define a custom interface to have this functionationality.

File: src/Handler/ExceptionHandler.php [hunk 17]
```diff
@@ -74,77 +83,116 @@ public function __invoke(
...
-            $handler = $this->handler($type);
-            $this->whoops->pushHandler($handler);
-
-            $body = $this->whoops->handleException($e);
-            $response->getBody()->write($body);
-
-            $this->whoops->popHandler();
-
-            return $response;
+        } catch (ExceptionInterface $exception) {
```
Comment (shadowhand) [inline src/Handler/ExceptionHandler.php]: I think I would rather see this implemented as an injection, where the ExceptionHandler takes a type of object that converts an exception into a Payload and then formats it using a Formatter.

The current implementation is not very flexible for user extension and doesn't allow for writing custom handlers based on 3rd party exception types.
